### ISSUE TITLE
Constants in patterns can contain equality-less types

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#34784]
Code:

``` rust
const C: *const str = "abcd";

fn main() {
    match C {
        C => {}
        _ => {}
    }
}
```

Diagnostics:

```
warning: broken MIR (Terminator { source_info: SourceInfo { span: <anon>:5:9: 5:10, scope: scope0 }, kind: tmp1 = <str as std::cmp::PartialEq>::eq(tmp0, const "abcd") -> bb3 }): bad arg #0 (&str <- *const str): Sorts(ExpectedFound { expected: &str, found: *const str })
 --> <anon>:4:11
  |>
4 |>     match C {
  |>           ^
```

Reproduces on stable/beta/nightly.

**Comment by eddyb** [comment]
This is a pointer pattern, it shouldn't compile AFAIK. cc @arielb1

**Comment by arielb1** [comment]
Yeah, this shouldn't really exist.

**Comment by nagisa** [comment]
As far as I remember, the constant-evaluated initializer expression gets used as a pattern, i.e. the code below and in the report are equivalent for MIR builder:

```rust
const C: *const str = "abcd";

fn main() {
    match C {
        "abcd" => {}
        _ => {}
    }
}
```

Which is caught by the proper-typeck and is why the mir-typeck is complaining.

**Comment by arielb1** [comment]
This is more general than raw pointers

```Rust
const C: fn() = main;

fn foo(x: fn()) {
    match x {
        C => {}
        _ => {}
    }
}

fn main() {}
```

However, other cases result in a const-eval error:

```Rust
const C: *const u8 = &0; //~ ERROR error: constant evaluation error: unimplemented constant expression: address operator [E0471]

fn foo(x: *const u8) {
    match x {
        C => {}
        _ => {}
    }
}

fn main() {}
```

I think this is the right solution.

**Comment by arielb1** [comment]
Problem line: https://github.com/rust-lang/rust/blob/1.10.0/src/librustc_const_eval/eval.rs#L340

**Comment by petrochenkov** [comment]
>@arielb1 changed the title from warning: broken MIR: expected: &str, found: *const str to Constants in patterns can contain equality-less types

Since you generalized this issue, https://github.com/rust-lang/rust/issues/34782 can probably be closed as a duplicate.

**Comment by arielb1** [comment]
cc @oli-obk

**Comment by petrochenkov** [comment]
One more test, for thin pointers:
```
const C: *const [u8; 4] = b"abcd";

fn main() {
    match C {
        C => {}
        _ => {}
    }
}
```
Gives `internal compiler error: ../src/librustc_trans/collector.rs:983: find_vtable_types_for_unsizing: invalid coercion *const [u8; 4] -> &[u8]` currently.

**Comment by oli-obk** [comment]
The issue is that the mir-building process calls `eval_const_expr` (https://github.com/rust-lang/rust/blob/f93aaf84cb50dfaaba44b08c05bd51320263f592/src/librustc_mir/hair/cx/pattern.rs#L68) while the old trans calls `consts::const_expr` (https://github.com/rust-lang/rust/blob/f93aaf84cb50dfaaba44b08c05bd51320263f592/src/librustc_trans/_match.rs#L288). The latter does things like unsizing and other coercion stuff.

It should work if instead of calling `eval_const_expr` we'd build a MIR for that expression and used the MIR-const evaluator on that MIR.

**Comment by arielb1** [comment]
@oli-obk

The problem is *that the code compiles*.
